### PR TITLE
Require `use` or `import` of outer scope before accessing symbols

### PR DESCRIPTION
**Pull request by bradcray**
This PR makes it so that a `use` or `require` is needed within a sub-module in order to access an outer scope's symbols (including sibling modules' symbols).  I was initially tepid on this idea, but have warmed up to it in part due to @ben-albrecht's argument in issue https://github.com/Cray/chapel-private/issues/884 and in part based on the experience of implementing it and seeing the required changes in our modules.

I was originally more in favor of requiring `use`/`import` of submodules than of parent modules because it seemed more natural to need to request access to name something nested below you in lexical scope, but not to have to do so for a higher scope.  For example, in:

```chapel
{
  var x = 1;
  {
    var y = 2;
    {
      var z = 3;
    }
  }
}
```

since `z`'s scope can "see" `x` and `y`, it seemed like an inner module scope should not need to do as much to refer to an outer module scope; and similarly since `x`'s scope can't see `z`, that an outer module scope would have to exert some effort to refer to an inner one's symbols.

It turns out that requiring such `use`/`import` statements (aside from the virtues Ben points out w.r.t. separate compilation) creates a fairly equivalent situation to what we already do for top-level modules.  Historically, we argued that `use` of a top-level module seemed unnecessary because lexically, you can "see it" already.  However, requiring such `use`s has turned out to make things more clear and explicit in our code while avoiding surprises.  Thus, it's now the case that:

```chapel
module A {
  // must `import B` before writing `B.x`
  var x = 42;
}
module B {
  // must `import A` before writing `A.x`
  var x = 33;
}
```

The same principle applies here, simply at the sub-module level:

```chapel
module Parent {
  module A {
    var x = 42;
  }
  module B {
    var x = 33;
  }
}
```

Traditionally, `A` could refer to `B.x` through normal lexical scoping, but with this PR it requires a `use`/`import` of `B`, analogously to the top-level module case.  

With this PR, `A` and `B` must also now `use`/`import` `Parent` to refer to its symbols which doesn't have an analogy in the top-level module world, but wasn't all that impactful to existing code, and adds clarity for submodules defined in separate files.

TODO (before merging):
- [ ] resolve a few tests that are still failing because they were designed to see what does / does not work, and I wasn't sure whether we should update the .goods now that they don't work, or update the tests to try and keep the behavior the same through additional `use`/`import` statements.

### CODE REVIEW

**Comment by bradcray** [comment]
Here are the tests that still need some updating based on whether we want to lock in the current errors or update the tests to keep the results the same or something else.  If we decide to proceed with this, I could use some help from @daviditen and @lydia-duncan with the ones they're familiar with:

[Error matching compiler output for modules/diten/usemodule/inner_uses_all_outer]
[Error matching compiler output for modules/diten/usemodule/test_nested_use_outer]
[Error matching compiler output for modules/diten/nested_module_collision]
[Error matching compiler output for modules/diten/nesting]
[Error matching compiler output for visibility/private/config/constants/privateToParent]
[Error matching compiler output for visibility/private/config/params/privateToParent]
[Error matching compiler output for visibility/private/config/variables/privateToParent]
[Error matching compiler output for visibility/private/constants/privateToParent]
[Error matching compiler output for visibility/private/moduleSymbols/accessPrivateSiblingSubmodule]
[Error matching compiler output for visibility/private/moduleSymbols/accessPrivateUsedSibling]
[Error matching compiler output for visibility/private/params/privateToParent]

File: modules/packages/Sort.chpl [hunk 1]
```diff
@@ -891,6 +896,8 @@ module MergeSort {
 
 pragma "no doc"
 module QuickSort {
+  import Sort.defaultComparator;
+  use super.ShallowCopy;
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I think I'd feel better if it was written `Sort.ShallowCopy`. Does that work?

File: test/modules/deitz/test_nested_modules2.chpl [hunk 2]
```diff
@@ -3,9 +3,10 @@ module OuterModule {
     var x: int = 2;
   }
 
-  use M1;
+  public use M1;
```
Comment (mppf) [inline test/modules/deitz/test_nested_modules2.chpl]: I don't think this should have to be 'public use'. If it doesn't work if it is not, let's add a future / issue about it.

File: test/modules/diten/usemodule/check_symbol_accesses.chpl [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
-  // "use outermost;".  check_symbol_accesses2 checks an illegal access.
+  // This test and check_symbol_accesses* test that the variable "a"
+  // can only be accessed by things that "use outermost;".
+  // check_symbol_accesses2 and 3 check an illegal access.
   module outermost {
     var a = 3;
     module middlemost {
       module innermost {
         proc foo() {
+	  import OuterModule.outermost.a;
```
Comment (mppf) [inline test/modules/diten/usemodule/check_symbol_accesses.chpl]: needs more indenting?

File: test/modules/diten/usemodule/check_symbol_accesses2.chpl [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
-  // "use outermost;".  check_symbol_accesses checks a legal access.
+  // This test and check_symbol_accesses* test that the variable "a"
+  // can only be accessed by things that "use outermost;".
+  // check_symbol_accesses checks a legal access and 3 checks another illegal.
   module outermost {
     var a = 3;
     module middlemost {
       module innermost {
         proc foo() {
+	  import OuterModule.outermost.a;
```
Comment (mppf) [inline test/modules/diten/usemodule/check_symbol_accesses2.chpl]: ditto

File: test/modules/deitz/test_nested_modules2.chpl (hunk 2 above)
Comment (bradcray) [reply test/modules/deitz/test_nested_modules2.chpl]: That doesn't seem right to me:  `use` is `private` / non-transitive by default, so this one permits OuterModule to refer to M1.x, but not for things that `use OuterModule` to do so.  I don't think we've discussed making inner modules any different w.r.t. these rules as other modules.

If M2 had a `use OuterModule.M1;` or `use super.M1;` then it wouldn't need to be public.

File: test/modules/diten/usemodule/check_symbol_accesses.chpl (hunk 3 above)
Comment (bradcray) [reply test/modules/diten/usemodule/check_symbol_accesses.chpl]: Needed a TAB replaced by spaces...  I'm still working through the hitches in getting my new home laptop in synch with my other ones.

File: test/modules/deitz/test_nested_modules2.chpl (hunk 2 above)
Comment (mppf) [reply test/modules/deitz/test_nested_modules2.chpl]: >  I don't think we've discussed making inner modules any different w.r.t. these rules as other modules

Right, I think we should discuss that.

Comment (bradcray) [reply test/modules/deitz/test_nested_modules2.chpl]: OK, I think you'd better write up the issue then.

File: modules/packages/Sort.chpl (hunk 1 above)
Comment (bradcray) [reply modules/packages/Sort.chpl]: Yep, done now.

File: test/modules/deitz/test_nested_modules2.chpl (hunk 2 above)
Comment (bradcray) [reply test/modules/deitz/test_nested_modules2.chpl]: Hmm, weird...  It seems we're inconsistent depending on whether the symbol in question is a function or a variable (or a mix).  I'm filing some futures in my PR about this.

Comment (bradcray) [reply test/modules/deitz/test_nested_modules2.chpl]: See https://github.com/chapel-lang/chapel/pull/15290/commits/e691c2c1b60209a737e0c4a00713133d792d4bf5

Comment (bradcray) [reply test/modules/deitz/test_nested_modules2.chpl]: Ohhhh... The reason we're inconsistent about functions and variables is that this PR's changes only affect scopeResolve.cpp and not function visibility.  That requires a symmetric effort.  Looking into it now, but won't be merging today.

**Comment by bradcray** [comment]
Closing in favor of #15312.

**Pull request closed without merging**